### DISCUSSION TITLE
Support for proxying requests via middleware

### DISCUSSION THREAD

**Discussion by TheBassGuy** [withastro/roadmap#665] [Proposal]
### Body

# Summary

Lets say I want to proxy `app.somewhere.com` to proxy `somewhere.com/app` so that I can display `somewhere.com/app` on `app.somewhere.com`

# Background & Motivation

NextResponse.rewrite() motivated me to request for this function, the above mentioned already exists in next js

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- Proxying requests
- Multi Tenant Apps

# Example

Mentioned above in summary...

**Comment by natemoo-re** [comment]
Currently this is possible by returning `fetch` to the destination, but you also need to add a header like `X-Astro-Rewrite` and check for it to guard against infinite redirects. Would be nice to have this built-in!

**Reply by TheBassGuy** [reply]
In-built would be amazing

**Reply by debasispanigrahi** [reply]
@natemoo-re can u please add an example for describing this process. Thanks in advance

**Reply by TheBassGuy** [reply]
agreed.

**Reply by TheBassGuy** [reply]
> Currently this is possible by returning `fetch` to the destination, but you also need to add a header like `X-Astro-Rewrite` and check for it to guard against infinite redirects. Would be nice to have this built-in!

Hi

I tried something similar, can't figure out in which request to check the header for, also when I do this whenever there is a case of 4XX series response code (as I tried) there is an infinite loop somewhere only in case of a 200 it does not loop over. Please can u send a middleware.ts file where I can see how this is done, here is the link of a project in which I tried implementing this - https://github.com/TheBassGuy/astro-multitenant-middleware

Thanks

**Reply by TheBassGuy** [reply]
I have implemented a check for context.response.ok, but that approach doesn't seem very good and only 2XX series responses are forwarded, all other responses are redirected to a 404 page

**Comment by lilnasy** [comment]
Wrote a tentative RFC here
https://github.com/withastro/roadmap/discussions/681

**Comment by u3u** [comment]
The [Rewriting to an external URL](https://nextjs.org/docs/app/api-reference/next-config-js/rewrites#rewriting-to-an-external-url) feature of Next.js has been very useful for me. I hope we can [Incremental adoption of Astro](https://nextjs.org/docs/app/api-reference/next-config-js/rewrites#incremental-adoption-of-nextjs) as well. 
Is there any other solution now?

**Reply by lilnasy** [reply]
That's a good use case. Currently, you would use a reverse proxy (nginx, traeffik, envoy) and it can forward the request to either your Astro app or your other app based on some detail about the request.

**Reply by u3u** [reply]
Yes, currently it seems that the only solution is to use Nginx. However, if there are many pages, I need to configure all the pages that are not in Astro's routing once, which is cumbersome and cannot be consistent with the development environment using `astro dev`. It can only be viewed in production for effects.

**Reply by lilnasy** [reply]
It's a bit more advanced, but it would be possible by changing a few lines in the adapter you are using.

FWIW, I will be recommending stage 2 as soon as I try out the @vercel/edge approach mentioned in the RFC I linked above. It may not be long before this proposal moves forward.

**Reply by u3u** [reply]
Hi @lilnasy , do you know how to change the adapter to implement it? Can you provide an example, thank you!

**Comment by u3u** [comment]
I am trying to solve the gradual adoption by adding a middleware and an empty low-priority dynamic route `pages/[...path].astro` (otherwise, 404 will not enter the middleware).

Due to the nature of route priority, named routes or nested dynamic routes defined in Astro will override it. The only way I can think of now is to determine if the middleware is handling the `[...path].astro` route by checking if the response text is "empty". It would be even better if we could get the current handled route name from the middleware context, such as `context.pathname` returning `/[...path]`. I have tested that this approach works fine, and all resources from external websites can be displayed correctly in Astro.

```ts
import { defineMiddleware } from 'astro:middleware';

export const onRequest = defineMiddleware(async ({ url, request }, next) => {
  const response = await next();
  const html: string = await response.text();
  const isFallback = html === '<!DOCTYPE html>'; // check is `/[...path].astro` ?

  if (isFallback && !request.headers.has('X-Astro-Rewrite')) {
    return fetch(`https://example.com${url.pathname}${url.search}`, {
      ...request,
      headers: {
        ...request.headers,
        'X-Astro-Rewrite': 'true',
      },
    });
  }

  return new Response(html, response);
});

```

**Reply by u3u** [reply]
Before an official solution is available, this solution is the best I can think of. There is only one situation where it may cause a problem, which is if there is a completely blank route in your Astro, but that's not possible 🤣

**Reply by lilnasy** [reply]
I don't think this needs a proposal. You can use an API Route and fetch within it.

```ts
// src/pages/legacy-route.ts

export const GET = ctx => fetch(`https://example.com/legacy-route`)
```

**Reply by u3u** [reply]
😮 Thank you! I forgot to have this! API endpoints also support dynamic routing, which is enough 👍

**`pages/[...path].ts`**

```ts
import type { APIRoute } from 'astro';

export const GET: APIRoute = ({ url }) => {
  return fetch(`https://example.com${url.pathname}${url.search}`);
};

```

**Reply by smnbbrv** [reply]
For such migrations I earlier used undici lib with connection pooling and real streaming, which is now open-sourced in the package https://github.com/smnbbrv/astroscope/tree/main/packages/proxy . Designed for SSR and is way more prepared for production as just fetch.
